File: maui/Services/SarAwaiter.cs
using Android.App;
using Android.Content;
using System;
using System.Threading.Tasks;

namespace fiskaltrust.Middleware.Demo
{
    static class SarAwaiter
    {
        static int _next = 1000;
        static readonly System.Collections.Concurrent.ConcurrentDictionary<int,
            TaskCompletionSource<Intent>> _pending = new System.Collections.Concurrent.ConcurrentDictionary<int,
            TaskCompletionSource<Intent>>();

        public static Task<Intent> StartForResultAsync(this Activity a, Intent i)
        {
            var rc = System.Threading.Interlocked.Increment(ref _next);
            var tcs = new TaskCompletionSource<Intent>();
            _pending[rc] = tcs;
            a.StartActivityForResult(i, rc);
            return tcs.Task;
        }

        public static bool Complete(int rc, Result code, Intent? data)
        {
            if (_pending.TryRemove(rc, out var tcs))
            {
                if (code == Result.Ok && data != null) tcs.TrySetResult(data);
                else tcs.TrySetException(new OperationCanceledException());
                return true;
            }
            return false;
        }
    }
}

File: java/app/src/main/java/eu/fiskaltrust/middleware/demo/SarAwaiter.java
package eu.fiskaltrust.middleware.demo;

import android.app.Activity;
import android.content.Intent;

import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.atomic.AtomicInteger;

public final class SarAwaiter {

  public interface ResultListener {
    void onResult(int resultCode, Intent data);
  }

  private static final AtomicInteger nextRequestCode = new AtomicInteger(2000);
  private static final ConcurrentHashMap<Integer, ResultListener> pending = new ConcurrentHashMap<>();

  private SarAwaiter() { }

  public static void startForResult(Activity activity, Intent intent, ResultListener listener) {
    int requestCode = nextRequestCode.incrementAndGet();
    pending.put(requestCode, listener);
    activity.startActivityForResult(intent, requestCode);
  }

  static boolean deliverResult(int requestCode, int resultCode, Intent data) {
    ResultListener listener = pending.remove(requestCode);
    if (listener == null) {
      return false;
    }
    listener.onResult(resultCode, data);
    return true;
  }
}
